File: csrc/src/top_k_per_row_prefill.cpp
/**
 * Copyright 2025 Enflame. All Rights Reserved.
 */
#include "top_k_per_row_prefill.h"

#include <topsaten/topsaten_vllm.h>
#include <torch/all.h>

#include "tops_extension/tops/Context.h"
#include "tops_extension/torch/GCUAten.h"
#include "torch_gcu.h"

namespace vllm_gcu::llm_ops {

void top_k_per_row_prefill(const at::Tensor& logits,
                           const at::Tensor& rowStarts,
                           const at::Tensor& rowEnds, at::Tensor& indices,
                           int64_t numRows, int64_t stride0, int64_t stride1,
                           int64_t topK) {
  const torch_gcu::OptionalGCUGuard device_guard(device_of(logits));
  const topsStream_t stream = torch_gcu::getCurrentGCUStream();
  if (numRows == 0) {
    return;
  }
  ATEN_ATENOP_CHECK(ATEN_ATENOP_CALL(topsvllm::topsvllmTopkPerRowPrefill)(
      indices, logits, rowStarts, rowEnds, numRows, stride0, stride1, topK, -1,
      stream));
}

}  // namespace vllm_gcu::llm_ops

File: vllm_gcu/kernels/native_op/top_k_per_row_prefill.py
import torch
from vllm_gcu.kernels.native_op.utils import register_native

@register_native("_C", "top_k_per_row_prefill")
def _ref_top_k_per_row_prefill(
    logits: torch.Tensor,
    row_starts: torch.Tensor,
    row_ends: torch.Tensor,
    indices: torch.Tensor,
    num_rows: int,
    stride0: int,
    stride1: int,
    topk: int,
    threshold:float
):
    topk_indices = logits.topk(min(topk, logits.shape[-1]),
                                dim=-1)[1]
    topk_indices -= row_starts[:, None]
    mask_lo = topk_indices >= 0
    mask_hi = topk_indices - (row_ends -
                                row_starts)[:, None] < 0
    mask = torch.full_like(topk_indices,
                            False,
                            dtype=torch.bool,
                            device=topk_indices.device)
    mask = mask_lo & mask_hi
    topk_indices = topk_indices.masked_fill(~mask, -1)
    indices[:, :topk_indices.shape[1]] = topk_indices.to(dtype=torch.int32)

